osbuild/pipeline.py

```diff
@@ -344,6 +356,55 @@ def download(self, store, monitor, libdir):
             for source in self.sources:
                 source.download(mgr, store, libdir)
 
+    def depsolve(self, store, targets: Iterable[str]) -> List[str]:
+        """Return the list of pipelines that need to be built
+
+        Given a list of target pipelines, return the names
+        of all pipelines and their dependencies that are not
+        already present in the store.
+        """
+
+        # A stack of pipelines to check if they need to be built
+        check = list(map(self.get, targets))
```

Comment:
I don't think this is a bug, but given that we consume `check` backwards, maybe we should reverse the passed in list here? For example, if you pass in  `["a", "b"]` it will currently return `["bdep", "b", "adep", "a"]`, but maybe `["adep", "a", "bdep", "b"]` is less surprising?

Reply:
Good point!

Reply:
I had a look and I think we already do the right thing; assuming `targets` is `["a", "b"]` (with no dependencies):

`check` will be `["a", "b"]`, but we consume it in reverse order, i.e. as `["b", "a"]`, thus
`build` will be `["b", "a"]`, since we do `build.move_to_end(<x>)`, thus we preserve the order of `check`, and so
result will be `["a", "b"]`, since we do `reversed(build)`.

With `a` and `b` having `a-dep` and `b-dep` as respective dependencies we will get:
- `check`: `["a", "b"]`, check`.pop()` (popped: `"b"`) gives
- `check`: `["a"]`, `check.append("b-dep")` (the dependency of b) gives
- `check`: `["a", "b-dep"]`, `check.pop()` (popped: `"b-dep"`) gives
- `check`: `["a"]`, `check.pop()` (popped: `"a"`) gives
- `check`: `[]`, `check.append("a-dep")` (the dependency of `"a"`)
- `check`: `["a-dep"]`, `check.pop()` (popped: `"a-dep"`) gives
- `check`: `[]` and here we terminate

so effectively we consumed `check` as `["b", "b-dep", "a", "a-dep"]` leading to `build` having the same order and thus we get the result, after reversal, `["a-dep", "a", "b-dep", "b"]`.

I did add a test (locally, not pushed yet):
```
       res = manifest.depsolve(store, names(ul, image))
       assert res == names(build, dep, ul, rootfs, image)
```
And that is already passing, unmodified.